zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/HDFSNotebookRepo.java

```diff
@@ -0,0 +1,503 @@
+          scheduledExecutorService
+              .schedule(this, getKerberosRefreshInterval(), TimeUnit.MILLISECONDS);
+        } else {
+          kinitFailCount++;
+          LOG.info("runKerberosLogin failed for " + kinitFailCount + " time(s).");
+          // schedule another retry at once or close the interpreter if too many times kinit fails
+          if (kinitFailCount >= kinitFailThreshold()) {
+            LOG.error("runKerberosLogin failed for  max attempts, calling close interpreter.");
+            close();
+          } else {
+            scheduledExecutorService.submit(this);
+          }
+        }
+        return null;
+      }
+    });
+
+    return scheduledExecutorService;
+  }
+
+  protected boolean isKerboseEnabled() {
+    if (!StringUtils.isEmpty(conf.getKerberosPrincipal()) &&
+        !StringUtils.isEmpty(conf.getKerberoskeyTab())) {
+      return true;
+    }
+    return false;
+  }
+
+  protected boolean runKerberosLogin() {
+    try {
+      Configuration conf = new
+          org.apache.hadoop.conf.Configuration();
+      conf.set("hadoop.security.authentication", KERBEROS.toString());
+      UserGroupInformation.setConfiguration(conf);
+      try {
+        UserGroupInformation.loginUserFromKeytab(
+            this.conf.getKerberosPrincipal(),
+            this.conf.getKerberoskeyTab()
+        );
+
+        if (UserGroupInformation.isLoginKeytabBased()) {
+          UserGroupInformation.getLoginUser().reloginFromKeytab();
+          return true;
+        } else if (UserGroupInformation.isLoginTicketBased()) {
+          UserGroupInformation.getLoginUser().reloginFromTicketCache();
+          return true;
+        }
+
+      } catch (IOException e) {
+        LOG.error("Failed to get either keytab location or principal name in the " +
+            "interpreter", e);
+      }
+      return true;
+    } catch (Exception e) {
+      LOG.error("Unable to run kinit for zeppelin", e);
+    }
+    return false;
+  }
+
+  private void setNotebookDirectory(String notebookDirPath) throws IOException {
+    stringPath = notebookDirPath;
+    setNotebookDirPath(getPath(stringPath));
+    if (!fileSystem.exists(getRootDir())) {
+      LOG.info("Notebook dir doesn't exist, create on is {}.", stringPath);
+      createFolder(this.notebookDirPath);
+    }
+  }
+
+  protected Path getRootDir() throws IOException {
+    if (!fileSystem.exists(fileSystem.getHomeDirectory())) {
+      throw new IOException("Root path does not exists");
+    }
+    if (!isDirectory(fileSystem.getHomeDirectory())) {
+      throw new IOException("Root path is not a directory");
+    }
+    return getNotebookDirPath();
+  }
+
+  private Path getPath(String path) {
+    String absolutePath;
+    if (path == null || path.trim().length() == 0) {
+      absolutePath = fileSystem.getHomeDirectory().toString();
+    } else if (path.startsWith("/")) {
+      absolutePath = fileSystem.getHomeDirectory().toString() + path;
```

Comment:
if a HDFS path starts with `/`, eg. `/root` - shouldn't the absolutePath be `/root`?